doc/src/manual/functions.md

```diff
@@ -66,16 +66,23 @@ Python, Ruby and Perl, among other dynamic languages.
 ## Argument-type declarations
 
 You can declare the types of function arguments by appending `::TypeName` to the argument name, as usual for [Type Declarations](@ref) in Julia.
-For example, the following function computes [Fibonacci numbers](https://en.wikipedia.org/wiki/Fibonacci_number) recursively:
+For example, the following function computes [Fibonacci numbers](https://en.wikipedia.org/wiki/Fibonacci_number) and returns the `n`th fibonacci number:
 ```
-fib(n::Integer) = n ≤ 2 ? one(n) : fib(n-1) + fib(n-2)
+julia> function fib(n::Integer)
+           a, b = 0, 1
+           while n > 1
+               a, b = b, a+b
+               n -= 1
+           end
+           a
+       end
```

Comment:
This function definition is incorrect. For example, it returns 34 as `fib(10)`. Please use any of the versions below. To my mind, `fib_w2` is the most similar to the recursive version, but it is still a linear iterative process, not a tree-recursive process.
```julia
# The linear iterative process
function fib(n, a=zero(n), b=one(n))
    for _ = 1:n
        b′ = b + a
        a = b
        b = b′
    end
    a
end
# Using a while loop to express linear iterative process
function fib_w(n, a=zero(n), b=one(n))
    m = 0
    while m < n
        b′ = b + a
        a = b
        b = b′
        m += 1
    end
    a
end
# Another while loop, decrementing counter rather than incrementing --
# less intuitive given the manner in which the computational process proceeds
function fib_w2(n, a=zero(n), b=one(n))
    while n > 0
        b′ = b + a
        a = b
        b = b′
        n -= 1
    end
    a
end
```

Reply:
These all seem much more opaque than the recursive definition, which is a fairly literal translation of the definition of Fibonacci you learn in kindergarten. 

If the recursion is really too complicated, an entirely different function could be used. But is there really a problem to solve here in the first place, at all?

Reply:
Why not use the simplest function possible? Something like `plus_one(n::Integer) = n + 1`

Reply:
@andrewjradcliffe I do not know what you mean by saying that definition is incorrect. What is `fib(10)` suppose to return when you start the fibonacci counting from `0` (which is the right thing).

Moreover the fact that people are obsessed about the right version of the fibonacci function tells you how we care about making things unnecessarily complex for the "naive users".

@andrewjradcliffe the function you introduced are very good but then of what use will it be for users are who hearing about functions for the first time and do not know about **"keyword arguments"**, the `one` function, the `zero` function. So you see, there's so much unnecessarily complex stuffs being introduced there.

My point is, that part of the doc is introducing functions, so don't make it any complex for anyone who only knows about `while` loop.

Reply:
My position is that the original (recursive) definition is completely fine (and, moreover, is preferable). It is good to see that others agree that the loop versions introduce unnecessary complication.

Should not `fib(10)=55`? The `fib` function as given in OP evaluates to
```julia
julia> fib.(0:10)
11-element Vector{Int64}:
  0
  0
  1
  1
  2
  3
  5
  8
 13
 21
 34
```

Reply:
From [wikipedia on Fibonacci numbers](https://en.wikipedia.org/wiki/Fibonacci_number#Binet%27s_formula), counting starts at `0` and then the next is `1`, so I don't see how `fib(10)` should return `55`. `fib.(0:10)` gives you that result because it returns the `n`th Fibonacci number, so `fib(1)` is still `0` just as `fib(0)` is `0`. IMO, this shouldn't even be the main point of the PR (anyone can implement the best `fib` function and provide a suggestion on it here). The main point is this: COMPLEXITY ON THE RECURSIVE FUNCTION.

You said: "others agree that the loop versions introduce unnecssary complication". Well by "others", you should know you're talking of expert Julia users, which this [function section](https://docs.julialang.org/en/v1/manual/functions/#Argument-type-declarations) regarding this PR wasn't necessarily written for. That section was written for beginners and I don't see how in the world a `while` loop is complicated compared to a recursive function for a beginner.

I teach Julia a lot, and I opened that PR regarding the complexity it introduced to the students I was teaching. Apparently the `while` loop was easier to comprehend and follow with the lesson; I've never met a beginner who said "recursive functions" are less complicated to `while` loops; do they even know what's a recursive function in the first place?😳

Reply:
The version of `fib` in the manual returns Fₙ as shown on the wikipedia page, i.e. that `fib(1)=1`. This PR assigns a different meaning which departs from the wikipedia page. Consequently, the sequence generated is not the canonical 0,1,1,2,3,5,..., but instead inserts an extra 0 at the beginning, due to the convention it introduces. This arises from the re-definition of what `fib` computes, which is not Fₙ, but where `n` is re-defined as the respective index in the Fibonacci sequence. 

Thank you for explaining the origins. However, my conclusion is that the indirection introduced by this PR would make this section of the manual more involved than a function that simply returns Fₙ, matching the canonical sequence.

Reply:
https://www.youtube.com/watch?v=ghxQA3vvhsk&t=101s

Reply:
The `fib` as been updated to do what the docs `fib` does.  So `fib.(1:10)` returns:
```julia
julia> fibₙ.(1:10)
10-element Vector{Int64}:
  1
  1
  2
...
 55
```

Like I said earlier, this isn't the main goal of the PR, as anyone can compute a better `fib` and provide a suggestion for it here. The main goal of the PR is to make that part easy to follow for a beginner (who it was originally written for).

I gave earlier example of how something of a similar nature was done in another language (no comparisons), and it clearly shows where I'm driving at. Since that's for beginners in the language, any reason for introducing the complexity of a RECURSIVE FUNCTION?

Reply:
This PR itself invokes an appeal to popular opinion, the veracity of which is difficult to ascertain. Furthermore, the repeated use of appeal to emotion in order to justify the proposed change is not a helpful path.

There is a logical reason why the `while` loop version should be avoided: up to this point in a linear reading of the manual, control flow constructs are yet to be introduced. This is very likely intentional. Consider a complete novice -- what basis would there be to believe that they would understand the termination criterion of a while loop? (e.g. is the condition evaluated at the entrance or exit of the loop?) A problem as simple as the scoping of `n` in the loop is a valid point of confusion for a complete novice. 
Recursion avoids these subtle questions, as the entry/exit points of the function provide clear demarcations.

My position remains that the original version serves purpose quite well and does not need replacement. I fail to understand how recursion is such a big deal, even for a complete novice.